src/Npgsql/TaskTimeoutAndCancellation.cs

```diff
@@ -0,0 +1,134 @@
+        /// <typeparam name="TResult">The result <see cref="Type"/>.</typeparam>
+        /// <returns>The <see cref="Task{TResult}"/> representing the asynchronous wait.</returns>
+        internal static async Task<TResult> WaitAsync<TResult>(Func<CancellationToken, Task<TResult>> getTaskFunc, NpgsqlTimeout timeout, CancellationToken cancellationToken, bool useLegacyImplementation = false)
+        {
+            Task<TResult>? task = default;
+            await WaitAsync(ct => (Task)(task = getTaskFunc(ct)), timeout, cancellationToken, useLegacyImplementation);
+            return await task!;
+        }
+
+        /// <summary>
+        /// Utility that executes a non-cancellable task while allowing to timeout and/or cancel awaiting for it.
+        /// If the given task does not complete within <paramref name="timeout"/>, a <see cref="TimeoutException"/> is thrown.
+        /// </summary>
+        /// <param name="getTaskFunc">Gets a <see cref="Task"/>.</param>
+        /// <param name="timeout">The timeout after which the <see cref="Task"/> should be faulted with a <see cref="TimeoutException"/> if it hasn't otherwise completed.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for a cancellation request.</param>
+        /// <param name="useLegacyImplementation">An optional <see cref="Boolean"/> parameter for testing only. If set to true the method uses pre-dotnet 6.0 implementation for cancellation and timeout enforcement.</param>
+        /// <returns>The <see cref="Task"/> representing the asynchronous wait.</returns>
+        internal static async Task WaitAsync(Func<CancellationToken, Task> getTaskFunc, NpgsqlTimeout timeout, CancellationToken cancellationToken, bool useLegacyImplementation = false)
+        { 
+            Task? task = default;
+            using var combinedCts = timeout.IsSet ? CancellationTokenSource.CreateLinkedTokenSource(cancellationToken) : null;
+            try
+            {
+                try
+                {
+                    task = getTaskFunc(combinedCts?.Token ?? cancellationToken);
+                    await WithCancellationAndTimeoutLocal(task);
+                }
+                catch (TimeoutException) when (!task!.IsCompleted)
+                {
+                    // Attempt to stop the Task in progress.
+                    combinedCts?.Cancel();
+                    throw;
+                }
+            }
+            catch
+            {
+                // Prevent unobserved Task notifications by observing the failed Task exception.
+                // To test: comment the next line out and re-run TaskExtensionsTest.DelayedFaultedTaskCancellation.
+                _ = task!.ContinueWith( t => _ = t.Exception, CancellationToken.None, TaskContinuationOptions.OnlyOnFaulted, TaskScheduler.Current);
```

Comment:
In some cases, `getTaskFunc` can throw an exception, and not a task wrapping an exception (this can happen if the exception is thrown from a method not marked as `async`), which will lead to an nre here. We should make a continuation only after we've made sure the task isn't `null`.

Reply:
Good catch. Should just take `getTaskFunc` outside of try/catch:
```
            task = getTaskFunc(combinedCts?.Token ?? cancellationToken);
            try
            {
                    ...
            }
            catch ....
```

Reply:
I have posted the code update.